device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenProvidedAuthenticationProvider.java

```diff
@@ -0,0 +1,68 @@
+/*
+*  Copyright (c) Microsoft. All rights reserved.
+*  Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+package com.microsoft.azure.sdk.iot.device.auth;
+
+import com.microsoft.azure.sdk.iot.device.SasTokenProvider;
+
+import javax.net.ssl.SSLContext;
+
+public class IotHubSasTokenProvidedAuthenticationProvider extends IotHubSasTokenAuthenticationProvider
+{
+    SasTokenProvider sasTokenProvider;
+    char[] lastSasToken;
+
+    public IotHubSasTokenProvidedAuthenticationProvider(String hostName, String deviceId, String moduleId, SasTokenProvider sasTokenProvider, SSLContext sslContext) {
+        super(hostName, null, deviceId, moduleId, sslContext);
+
+        if (sasTokenProvider == null)
+        {
+            throw new IllegalArgumentException("sas token provider cannot be null");
+        }
+
+        this.sasTokenProvider = sasTokenProvider;
+    }
+
+    @Override
+    public boolean isRenewalNecessary()
+    {
+        return false;
+    }
+
+    @Override
+    public void setTokenValidSecs(long tokenValidSecs)
+    {
+        throw new UnsupportedOperationException("Cannot configure sas token time to live when custom sas token provider is in use");
+    }
+
+    @Override
+    public boolean canRefreshToken()
+    {
+        return true;
+    }
+
+    @Override
+    public char[] getSasToken()
+    {
+        lastSasToken = sasTokenProvider.getSasToken();
+        return lastSasToken;
+    }
+
+    @Override
+    public int getMillisecondsBeforeProactiveRenewal()
+    {
+        // Seconds since UNIX epoch when this sas token will expire
+        long expiryTimeSeconds = IotHubSasToken.getExpiryTimeFromToken(new String(lastSasToken));
+
+        // Assuming that the token's life "starts" now for the sake of figuring out when it needs to be renewed. Users
+        // could theoretically give us a SAS token that started a while ago, but since we have no way of figuring that out,
+        // we will conservatively just renew at 85% of the remaining time on the token, rather than 85% of the time the token
```

Comment:
I don't see any code referring to 85%

Reply:
That sits in a class that this class extends

Reply:
Seems all the talk about 85% then is irrelevant there. That responsibility is delegated to the base class. If it changes in the base class, this comment would be wrong, so best to just abstract it and say, the base class schedules early renewal via `this.timeBufferPercentage`, and we'll adhere to that in this impl.

Reply:
Good idea